Signal/src/ViewControllers/NewGroupViewController.m

```diff
@@ -421,39 +454,111 @@ - (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger
+    ContactTableViewCell *cell
+        = (ContactTableViewCell *)[tableView dequeueReusableCellWithIdentifier:[ContactTableViewCell reuseIdentifier]];
+    if (!cell) {
+        cell = [ContactTableViewCell new];
     }
 
-    NSUInteger row   = (NSUInteger)indexPath.row;
-    Contact *contact = contacts[row];
-
-    cell.textLabel.attributedText = [self.contactsManager formattedFullNameForContact:contact font:cell.textLabel.font];
+    [self updateContentsOfCell:cell indexPath:indexPath];
 
-    tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
+    return cell;
+}
 
-    if ([[tableView indexPathsForSelectedRows] containsObject:indexPath]) {
-        [self adjustSelected:cell];
+- (void)updateContentsOfCell:(ContactTableViewCell *)cell indexPath:(NSIndexPath *)indexPath
+{
+    OWSAssert(cell);
+    OWSAssert(indexPath);
+
+    Contact *contact = contacts[(NSUInteger)indexPath.row];
+    BOOL isBlocked = [self isContactBlocked:contact];
+    BOOL isInGroup = [self isContactInGroup:contact];
+    BOOL isSelected = [[self.tableView indexPathsForSelectedRows] containsObject:indexPath];
+    // More than one of these conditions might be true.
+    // In order of priority...
+    cell.accessoryMessage = nil;
+    cell.accessoryView = nil;
+    cell.accessoryType = UITableViewCellAccessoryNone;
+    if (isInGroup) {
+        OWSAssert(!isSelected);
+        // ...if the user is already in the group, indicate that.
+        cell.accessoryMessage = NSLocalizedString(
+            @"CONTACT_CELL_IS_IN_GROUP", @"An indicator that a contact is a member of the current group.");
+    } else if (isSelected) {
+        // ...if the user is being added to the group, indicate that.
+        cell.accessoryType = UITableViewCellAccessoryCheckmark;
+    } else if (isBlocked) {
+        // ...if the user is blocked, indicate that.
+        cell.accessoryMessage
+            = NSLocalizedString(@"CONTACT_CELL_IS_BLOCKED", @"An indicator that a contact has been blocked.");
     }
+    [cell configureWithContact:contact contactsManager:self.contactsManager];
+}
 
-    return cell;
+- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    return [ContactTableViewCell rowHeight];
 }
 
 #pragma mark - Table View delegate
+
 - (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
-    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
-    [self adjustSelected:cell];
-}
+    Contact *contact = contacts[(NSUInteger)indexPath.row];
+    BOOL isBlocked = [self isContactBlocked:contact];
+    BOOL isInGroup = [self isContactInGroup:contact];
+    if (isInGroup) {
+        // Deselect.
+        [tableView deselectRowAtIndexPath:indexPath animated:YES];
+
+        NSString *displayName = [_contactsManager displayNameForContact:contact];
+        UIAlertController *controller = [UIAlertController
```

Comment:
just a thought:

I haven't run this yet, but is it possible or likely that the user is tapping on an *existing* member of the group in order to "deselect" them aka to "kick" them from the group?

In that case, maybe we should phrase the copy something like:

```
title: %@ is already in the group. 
body: You cannot kick people from a group. Either they have to leave, or you can start a new group without this user.
```

Pretty verbose, I know. =/

thoughts?